### PR TITLE
Building manylinux2014 bdist_wheel

### PR DESCRIPTION
**Pull request by cheng-chi**
Since issues are not available in this repo, I will use this PR as a platform of discussion. The goal can be divided into three stages:
* Short term (1-2 weeks): Make useable python3.6, 3.7 manylinux bdist available on [PyPI](https://pypi.org/project/s2geometry/).
* Medium term: Have a reliable procedure to build bdist for future python versions.
* Long term: Build bdist on travis-ci, and merge changes into [google's original repo](https://github.com/google/s2geometry).

### CODE REVIEW

**Comment by cheng-chi** [comment]
As shown in docker/manylinux2014_s2geometry/Dockerfile, I have created a primitive build docker, available on [dockerhub](https://hub.docker.com/repository/docker/chicheng/manylinux2014_s2geometry). On top of the original [manylinux2014](https://github.com/pypa/manylinux) docker, this docker provides swig3, openssl, cmake and miniconda. Conda enviroments named "py35, py36, py37, py38" are also created. My current build process is similar to:
```
$ cd s2geometry
$ conda activate py36
$ python setup.py bdist_wheel
$ authwheel repair dist/s2geometry-0.9.0-cp36-cp36m-linux_x86_64.whl
$ cp wheelhouse/s2geometry-0.9.0-cp36-cp36m-manylinux2014_x86_64.whl ~
```
However, the wheel produced by this process has some problem. At import time, an import error will be raised:
```
Traceback (most recent call last):
  File "/root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/pywraps2.py", line 14, in swig_import_helper
    return importlib.import_module(mname)
  File "/root/miniconda3/envs/py37/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 670, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 583, in module_from_spec
  File "<frozen importlib._bootstrap_external>", line 1043, in create_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
ImportError: /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libcrypto-957ef781.so.1.0.2k: version `libcrypto.so.10' not found (required by /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/_pywraps2.so)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/__init__.py", line 1, in <module>
    from .pywraps2 import *
  File "/root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/pywraps2.py", line 17, in <module>
    _pywraps2 = swig_import_helper()
  File "/root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/pywraps2.py", line 16, in swig_import_helper
    return importlib.import_module('_pywraps2')
  File "/root/miniconda3/envs/py37/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
ModuleNotFoundError: No module named '_pywraps2'
```
Further examine of the _pywraps2.so in installed site-packages directory has shown that openssl's dynamic library can not be properly loaded. Note that auditwheel will collect needed dynamic libraries, rename and rewrite ELF's RPATH, SONAME, and then package these so files inside the new wheel. It seems that other people are having similiar issue as per this [issue](https://github.com/pypa/auditwheel/issues/217) on Auditwheel's repo.
```
$ ldd /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/_pywraps2.so
/root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/_pywraps2.so: /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libcrypto-957ef781.so.1.0.2k: version `libcrypto.so.10' not found (required by /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/_pywraps2.so)
/root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/_pywraps2.so: /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libcrypto-957ef781.so.1.0.2k: version `libcrypto.so.10' not found (required by /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libssl-84c8c57b.so.1.0.2k)
        linux-vdso.so.1 =>  (0x00007ffe42580000)
        libpython3.7m.so.1.0 => not found
        libssl-84c8c57b.so.1.0.2k => /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libssl-84c8c57b.so.1.0.2k (0x00007f5558779000)
        libcrypto-957ef781.so.1.0.2k => /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libcrypto-957ef781.so.1.0.2k (0x00007f55582e3000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f55580c7000)
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f5557dc0000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f5557abe000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f55578a8000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f55574da000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f5558e08000)
        libgssapi_krb5-92a3b421.so.2.2 => /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libgssapi_krb5-92a3b421.so.2.2 (0x00007f555727b000)
        libkrb5-c4b42758.so.3.3 => /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libkrb5-c4b42758.so.3.3 (0x00007f5556f77000)
        libcom_err-20a7370a.so.2.1 => /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libcom_err-20a7370a.so.2.1 (0x00007f5556d72000)
        libk5crypto-99d9cebb.so.3.1 => /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libk5crypto-99d9cebb.so.3.1 (0x00007f5556b3c000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f5556938000)
        libz-79aa6943.so.1.2.11 => /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libz-79aa6943.so.1.2.11 (0x00007f5558ffb000)
        libkrb5support-f5716915.so.0.1 => /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libkrb5support-f5716915.so.0.1 (0x00007f5556725000)
        libkeyutils-dfe70bd6.so.1.5 => /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libkeyutils-dfe70bd6.so.1.5 (0x00007f5556520000)
        libresolv.so.2 => /lib64/libresolv.so.2 (0x00007f5556307000)
        libselinux-8db37e25.so.1 => /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libselinux-8db37e25.so.1 (0x00007f55560db000)
        libpcre-9513aab5.so.1.2.0 => /root/miniconda3/envs/py37/lib/python3.7/site-packages/s2geometry/../s2geometry.libs/libpcre-9513aab5.so.1.2.0 (0x00007f5555e78000)
```

**Comment by cheng-chi** [comment]
As shown below, replacing the shared library name in _pywraps2.so with system openssl library name will solve this problem. But I'm not confident that other users will have the openssl version we need. In the meantime, I will investigate other options.
```
$readelf -d _pywraps2.so | head -15

Dynamic section at offset 0x224000 contains 33 entries:
  Tag        Type                         Name/Value
 0x000000000000000f (RPATH)              Library rpath: [$ORIGIN/../s2geometry.libs]
 0x0000000000000001 (NEEDED)             Shared library: [libpython3.7m.so.1.0]
 0x0000000000000001 (NEEDED)             Shared library: [libssl-84c8c57b.so.1.0.2k]
 0x0000000000000001 (NEEDED)             Shared library: [libcrypto-957ef781.so.1.0.2k]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
 0x000000000000000c (INIT)               0x4cb40
 0x000000000000000d (FINI)               0x1395b4
```
```
$ patchelf --replace-needed libcrypto-957ef781.so.1.0.2k libcrypto.so.1.0.2k _pywraps2.so
$ patchelf --replace-needed libssl-84c8c57b.so.1.0.2k libcrypto.so.1.0.2k _pywraps2.so
$ readelf -d _pywraps2.so | head -15

Dynamic section at offset 0x224000 contains 33 entries:
  Tag        Type                         Name/Value
 0x000000000000000f (RPATH)              Library rpath: [$ORIGIN/../s2geometry.libs]
 0x0000000000000001 (NEEDED)             Shared library: [libpython3.6m.so.1.0]
 0x0000000000000001 (NEEDED)             Shared library: [libssl.so.10]
 0x0000000000000001 (NEEDED)             Shared library: [libcrypto.so.10]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
 0x000000000000000c (INIT)               0x4cb40
 0x000000000000000d (FINI)               0x1395b4
```
```
$ ldd _pywraps2.so 
        linux-vdso.so.1 =>  (0x00007ffe4f972000)
        libpython3.6m.so.1.0 => not found
        libssl.so.1.0.2k => /lib64/libssl.so.1.0.2k (0x00007f79ec078000)
        libcrypto.so.1.0.2k => /lib64/libcrypto.so.1.0.2k (0x00007f79ebc15000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f79eb9f9000)
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f79eb6f2000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f79eb3f0000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f79eb1da000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f79eae0c000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f79ec76e000)
        libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2 (0x00007f79eabbf000)
        libkrb5.so.3 => /lib64/libkrb5.so.3 (0x00007f79ea8d6000)
        libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007f79ea6d2000)
        libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x00007f79ea49f000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f79ea29b000)
        libz.so.1 => /lib64/libz.so.1 (0x00007f79ea085000)
        libkrb5support.so.0 => /lib64/libkrb5support.so.0 (0x00007f79e9e75000)
        libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x00007f79e9c71000)
        libresolv.so.2 => /lib64/libresolv.so.2 (0x00007f79e9a58000)
        libselinux.so.1 => /lib64/libselinux.so.1 (0x00007f79e9831000)
        libpcre.so.1 => /lib64/libpcre.so.1 (0x00007f79e95cf000)
```

**Comment by cheng-chi** [comment]
One alternative is to statically link against openssl. @figroc Does this require major changes in the build scripts?

**Comment by figroc** [comment]
> One alternative is to statically link against openssl. @figroc Does this require major changes in the build scripts?

I didn't think about statically linking. It would be good if it could be done. The issue of auditwheel doesn't seem to be resolved in a short time.

File: docker/manylinux2014_s2geometry/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+# create python conda enviroments
+RUN eval "$(~/miniconda3/bin/conda shell.bash hook)" \
+  && conda create -n py35 python=3.5 -y \
+  && conda create -n py36 python=3.6 -y \
+  && conda create -n py37 python=3.7 -y \
+  && conda create -n py38 python=3.8 -y
+
+# default shell
+CMD ["/bin/bash"]
```
Comment (figroc) [inline docker/manylinux2014_s2geometry/Dockerfile]: Could you tweak the `build.sh` to use your docker image? The `Dockerfile` can be put under the `cross` dir in this repo,  or put in your repo and referenced in the `build.sh`. I am okay with either way.

**Comment by cheng-chi** [comment]
After statically link against openssl (as per f802035), the product of auditwheel are indeed useable. Tested for python3.5-3.8.

File: docker/manylinux2014_s2geometry/Dockerfile (hunk 1 above)
Comment (cheng-chi) [reply docker/manylinux2014_s2geometry/Dockerfile]: I will take a look.

**Comment by cheng-chi** [comment]
I have just updated the build.sh to use my dockerfile. It's a little ugly right now since it write everything back to local source directory. Please make changes as you see fit.

**Comment by cheng-chi** [comment]
@figroc Annnnnd is it possible to upload these bdist wheels to PyPI as soon as possible once you have tested them? I and other developers at Nuro kinda need it right now.

**Comment by figroc** [comment]
> @figroc Annnnnd is it possible to upload these bdist wheels to PyPI as soon as possible once you have tested them? I and other developers at Nuro kinda need it right now.

Yes, I will do that. After I tested your patch, I would upload the wheels. And patch your patch later.

**Comment by figroc** [comment]
`cross/test.sh` runs all green for manylinux2014 bdist_wheels. I've published them to PyPi. Thanks very much for your contribution.

**Comment by cheng-chi** [comment]
@figroc Thanks! I might work on building binary wheels for MacOS once I have time.

**Pull request merged**